Fixed #22828 -- Documented ModelAdmin getters returning class attributes.

Koressi opened the pull request:
#### Trac ticket number
<!-- Replace XXXXX with the corresponding Trac ticket number, or delete the line and write "N/A" if this is a trivial PR. -->

ticket-22828

#### Branch description
Add a warning inside ModelAdmin documentation about a bit obscure behaviour. All getters related to `ModelAdmin` properties return the properties themselves rather than a copy of their values, which can be misleading.

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [ ] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

sarahboyce reviewed (commented):
I think this warning should be broader than the specific example and any dynamic modifying should be discouraged

Koressi commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -1465,6 +1465,30 @@ templates used by the :class:`ModelAdmin` views:
+    the ``PersonAdmin.readonly_fields`` each time 
+    a non-superuser visits this admin page and you
+    might end up with ``readonly_fields = ["name", "age", "age", ...]``
+    even for a **superuser**.
```
Thank you for your time and the feedback! I absolutely agree with the changes you proposed. Should we still recommend a safer way to do that in this case? If so, would something like that work?
```suggestion 
A safer approach would be to work on a copy of the property.
Based on the example above::
    from copy import copy

    class PersonAdmin(admin.ModelAdmin):

        def get_readonly_fields(request, obj=None):
            readonly_base = super().get_readonly_fields(request, obj)
            readonly = copy(readonly_base)  # use deepcopy() for nested properties
            ...
```

Koressi commented:
Thank you for your time and the feedback! I absolutely agree with the changes you proposed. Should we still recommend a safer way to do that in this case? If so, would something like that work?
```suggestion 
A safer approach would be to work on a copy of the property.
Based on the example above::
    from copy import copy

    class PersonAdmin(admin.ModelAdmin):

        def get_readonly_fields(request, obj=None):
            readonly_base = super().get_readonly_fields(request, obj)
            readonly = copy(readonly_base)  # use deepcopy() for nested properties
            ...
```

sarahboyce commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -1465,6 +1465,27 @@ templates used by the :class:`ModelAdmin` views:
+.. warning::
+
+    All hooks that return a ``ModelAdmin`` property return the property itself
+    rather than a copy of their values. Dynamically modifying this value can
+    lead to surprising results.
+
+    Let's take :meth:`ModelAdmin.get_readonly_fields` as an example::
```
```suggestion
    rather than a copy of its value. Dynamically modifying this value can lead
    to surprising results. Let's take :meth:`ModelAdmin.get_readonly_fields` as
    an example::
```
(sorry I made a grammar mistake)

sarahboyce commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -1465,6 +1465,27 @@ templates used by the :class:`ModelAdmin` views:
+            def get_readonly_fields(request, obj=None):
+                readonly = super().get_readonly_fields(request, obj)
+                if not request.user.is_superuser:
+                    readonly.append("age")  # Edits the class attribute
+
+                return readonly
```
```suggestion
                    readonly.append("age")  # Edits the class attribute.
                return readonly
```

sarahboyce commented:
> A safer approach would be to work on a copy of the property.
> Based on the example above::
>     import copy
> 
>     class PersonAdmin(admin.ModelAdmin):
> 
>         def get_readonly_fields(request, obj=None):
>             readonly_base = super().get_readonly_fields(request, obj)
>             readonly = copy.deepcopy(readonly_base)
>             ...

I like it :star: 
I might go for `You should modify a copy instead::` rather than `A safer approach would be to work on a copy of the property. Based on the example above::` just as I'm conscious it's a long warning :thinking:

Koressi commented:
Yeah the warning got pretty long. I've added the recommended solution and incorporated the changes you proposed.

PS
I'm so sorry for all the formatting errors my pre-commit was acting up 😅

sarahboyce commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -1465,6 +1465,38 @@ templates used by the :class:`ModelAdmin` views:
+    Let's take :meth:`ModelAdmin.get_readonly_fields` as an example::
+
+        class PersonAdmin(admin.ModelAdmin):
+            readonly_fields = ["name"]
```
I added these back in, as after testing found that `readonly_fields` is an empty tuple by default (and therefore immutable) - sorry about that

sarahboyce commented:
I removed the last example as overriding the property not to be a tuple is a requirement and hence I feel this is more of a niche issue and wanted to reduce the size of the warning.

If you're happy with this, I will land this later this week

sarahboyce reviewed (approved):
Thank you @Koressi! Welcome onboard :boat:
